### PR TITLE
feat: add sample-level loss normalization for SFT

### PR DESCRIPTION
**Pull request by hallerite**
## Summary
- Adds `loss_normalization` config option (`"token"` | `"sample"`, default `"token"`) implementing per-conversation normalized loss (Nemotron 3 Super Stage 2, eq. 2)
- `"sample"` mode computes per-sample mean loss then averages across samples, preventing long outputs from dominating the gradient — useful for restoring long-input-short-output performance after token-level SFT
- Works with all packing modes (`cat`, `stack`), all loss implementations (`torch`, `liger`, `liger_fused`), and context parallelism

## How it works
- **`cat` packing**: infers sample boundaries from `position_ids` resets within packed sequences, uses `scatter_add` to group per-token losses by sample
- **`stack` packing**: each batch row is a sample, straightforward per-row normalization
- **`liger_fused`**: vendors Liger fused CE forward (~100 lines of Python, Triton kernel untouched) with a `token_weight` parameter that scales both per-token losses and pre-computed gradients during the chunked forward pass — preserves the memory savings of fused CE (no logit materialization)
- **Context parallelism**: each CP rank computes local per-sample loss sums, all-reduces per-sample token counts across the CP group so every rank divides by the correct global count

## Usage
```toml
loss_normalization = "sample"
```

## Memory comparison (Qwen3-0.6B, seq_len=4096, batch_size=4)
| | `torch` + sample | `liger_fused` + sample |
|--|--|--|
| Peak Memory | 23.3 GiB | 18.1 GiB |

5.2 GiB saved — fused kernel memory advantage fully preserved with sample normalization.

## Numerical verification
- `liger_fused` + `token` on branch is bit-identical at step 0 vs main, with normal float drift at later steps
- `torch` + `sample` vs `liger_fused` + `sample` produce matching loss trajectories (tiny float drift from different accumulation order)

## Test plan
- [x] 13 unit tests: config validation, boundary detection edge cases, both packing modes, masked tokens, gradient correctness
- [x] All 15 existing SFT dataset tests pass
- [x] 2-GPU: `torch` + `sample` + cat packing + variable-length data
- [x] 2-GPU: `torch` + `sample` + stack packing + variable-length data
- [x] 2-GPU: `liger_fused` + `sample` + real SFT data
- [x] 2-GPU: `liger_fused` + `token` regression check (bit-identical to main at step 0)
- [x] 2-GPU: CP=2 + `sample` + real SFT data
- [x] 2-GPU: CP=2 + `sample` + fixed-length fake data
- [x] Memory comparison: `torch` vs `liger_fused` with sample norm (5.2 GiB savings confirmed)

---

> [!NOTE]
> **Medium Risk**
> Changes SFT loss/gradient scaling semantics and adds a custom wrapper around Liger fused cross-entropy, which can affect training dynamics and correctness (especially with packing + context parallelism). Scope is contained to SFT training and LM-head loss computation paths.
> 
> **Overview**
> Adds `loss_normalization` (`token` default, `sample`) to `SFTConfig` and updates SFT training to optionally normalize loss per sample (mean per conversation) instead of globally per token.
> 
> Implements sample-boundary detection for `cat` packing via `position_ids` resets and computes per-sample means for both `cat` and `stack`, including CP-aware aggregation of per-sample token counts.
> 
> Extends the model/lm-head forward path with an optional `token_weight` and introduces a vendored Python wrapper (`fused_ce_sample_norm.py`) to apply per-token weights to Liger’s fused CE loss/gradients so sample-normalization can be used without materializing logits. Adds unit tests covering boundary detection, packing modes, masking, and gradient behavior.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 85bc65fd43871530b9d1ae3171d31c7746c60155. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

**Comment by hallerite** [comment]
did not see any benefit on ablations, so closing this for now
